aiosmtpd/smtp.py

```diff
@@ -535,19 +560,25 @@ def smtp_DATA(self, arg):
         received_data = EMPTYBYTES.join(data)
         if self._decode_data:
             received_data = received_data.decode('utf-8')
-        args = (self.peer, self.mailfrom, self.rcpttos, received_data)
+        self.envelope.content = received_data
+        method = getattr(self.event_handler, 'handle_DATA', None)
+        args = (self.session, self.envelope)
         kwargs = {}
-        if not self._decode_data:
-            kwargs = {
-                'mail_options': self.mail_options,
-                'rcpt_options': self.rcpt_options,
-                }
-        kwargs.update({'loop': self.loop})
-        if asyncio.iscoroutinefunction(self.event_handler.process_message):
-            status = yield from self.event_handler.process_message(
-                *args, **kwargs)
+        if method is None:
+            method = self.event_handler.process_message
+            args = (
+                self.session.peer,
+                self.envelope.mail_from,
+                self.envelope.rcpt_tos,
+                self.envelope.content)
+            if not self._decode_data:
+                kwargs = {
+                    'mail_options': self.envelope.mail_options,
+                    'rcpt_options': self.envelope.rcpt_options}
+        if asyncio.iscoroutinefunction(method):
+            status = yield from method(*args, **kwargs)
         else:
-            status = self.event_handler.process_message(*args, **kwargs)
+            status = method(*args, **kwargs)
```

Comment:
sorry, looks like I just dumb tonight. Now look on it and thinking what part of me I used to read your comments. I totally agree with #39, but looking for way to just merge it over my changes. But can`t understand where its branch is placed. What do you think about landin it over this PR? Look that I need to just do it by hands.